### PR TITLE
Add gitconfig page to dev docs

### PR DESCRIPTION
**Pull request by gemmaguest**
**Description of work.**

This PR adds a page to the dev docs with some suggested config for git. This is currently a draft and is very much open to discussion - please let me know if something should be added/removed. I'm open to adjusting e.g. the log formatting suggestions as well. It would be good to have a base set of aliases that most people are happy with so we have a common point of reference.

**To test:**

- Build the dev docs and check that the new page reads and looks ok and is linked from the git workflow page.
- Check that the suggested config options work / make sense.

*There is no associated issue.*

*This does not require release notes* because **it relates to developer documentation**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: dev-docs/source/GitConfig.rst [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+.. code-block:: sh
+
+   [pretty]
+           concise = "%C(yellow)%h%Creset %C(green)[%ar]%Creset %C(auto)%d%Creset %s %C(cyan)<%an>"
+           detail = "%C(yellow)commit %h%Creset%C(auto)%d%Creset%n%C(yellow)Parents: %p%Creset%n%C(cyan)Author: %an <%ae>%Creset%n%C(cyan)        %ai (%ar)%Creset%n%C(green)Commit: %cn <%ce>%Creset%n%C(green)        %ci (%cr)%Creset%n%n%w(79)%s%n%n%b"
+
+   [alias]
+           # Check out a pull request for testing - uncomment and replace <main-remote-name>
+           #test-pr = "!f() { git fetch <main-remote-name> +pull/$1/merge:pr/$1-merged && git checkout pr/$1-merged; }; f"
```
Comment (DanNixon) [inline dev-docs/source/GitConfig.rst:122]: I prefer my version of `test-pr` (https://github.com/DanNixon/dotfiles/blob/master/roles/git/files/gitconfig#L49) as you don't have to hard code the remote name, and as such can make better use of multiple remotes.

File: dev-docs/source/GitConfig.rst [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+Mantid Git Config
+=================
+
+.. contents:: Contents
+   :local:
+
+Summary
+-------
+
+This page describes some configuration that will help with your usage of Git on the Mantid project - some of which is strongly recommended. This should all be added to your ``~/.gitconfig`` file and will be global across all of your repositories - please sanity check these suggestions if you use git for other repositories.
```
Comment (DanNixon) [inline dev-docs/source/GitConfig.rst:13]:
```suggestion
This page describes some configuration that will help with your usage of Git - some of which is strongly recommended. This should all be added to your ``~/.gitconfig`` file and will be global across all of your repositories - please sanity check these suggestions if you use git for other repositories.
```

None of this has anything to do with Mantid.

**Comment by DanNixon** [comment]
See https://github.com/DanNixon/dotfiles/blob/master/roles/git/files/gitconfig#L43-L45
The push aliases are almost essential, the fetch alias is very useful if you use multiple remotes.

**Comment by gemmaguest** [comment]
> 
> 
> See [https://github.com/DanNixon/dotfiles/blob/master/roles/git/files/gitconfig#L43-L45](https://github.com/DanNixon/dotfiles/blob/master/roles/git/files/gitconfig?rgh-link-date=2020-08-20T14%3A00%3A02Z#L43-L45)
> The push aliases are almost essential, the fetch alias is very useful if you use multiple remotes.

I have `fetch.prune=true` so `fetch --prune` shouldn't be required. I'll include the others - probably as `pushu` and `pushfl` so they're not confused with pull.

**Comment by gemmaguest** [comment]
I have `pull.rebase=true` but wondering if `pull.ff=only` is better...

**Comment by DanNixon** [comment]
> I have `fetch.prune=true` so `fetch --prune` shouldn't be required

Ah yes, good point. My config is quite a number of years old now, I should really make a point to update things where possible.

**Comment by DanNixon** [comment]
> I have `pull.rebase=true` but wondering if `pull.ff=only` is better...

They are different, `pull.ff=only` will prevent you from pulling when anything other than a fast forward can be done. Typically you should never be in a situation where a merge or rebase is needed after pulling so only allowing fast forwards is sensible IMO.

So, yes, IMO `pull.ff=only` is better. Unless someone can think of a case that I have overlooked.

**Comment by gemmaguest** [comment]
Any thoughts on these? I find them useful but am worried about over-complicating things for new starters.
```
[core]
        # Specify your own global list of files/patterns that git should ignore
        excludesfile = ~/.gitignore
[blame]
        # This file allows you to specify commits that should be ignored in git-blame e.g. bulk changes
        ignoreRevsFile = ~/.git-blame-ignore-revs
[rerere]
        # Reuse Recorded conflict Resolutions
        enabled = true
        # Auto-stage files resolved by rerere
        autoupdate = true
```

**Comment by DanNixon** [comment]
> Any thoughts on these? I find them useful but am worried about over-complicating things for new starters.

Happy with all of those, the global excludes are general good practice, I'm on the fence about blame (I probably would not use it, but it could be useful), rerere can be enabled and explained later when it is actually needed.

On the subject of over complication, I'd actually be tempted to remove the section on including config fragments. I find the ability to do this often makes configuration more complex (for single user cases at least, where the only issue with a single config file is complexity/size).

**Comment by DanNixon** [comment]
One more thing that I just remembered is worth aliasing: `add --patch`. This makes you think about what you are staging and is essential to creating focused, atomic commits when you have a larger number of changes.

In all honesty, this should be the default way people stage files.

**Comment by gemmaguest** [comment]
Thanks for the input @DanNixon. I've added an alias for `commit --amend`, too, which seems essential if you commit frequently and incrementally.

I'm leaving this as draft because I'd like to see if @martyngigg or @DavidFair want to contribute.

File: dev-docs/source/GitConfig.rst [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+   [alias]
+           # Check out a pull request for testing - uncomment and replace <main-remote-name>
+           test-pr = "!f() { git fetch $1 pull/$2/merge:pr/$2-merged && git checkout pr/$2-merged; }; f"
+           # Remove all branches starting pr/
+           test-pr-remove-all = "!f() { git branch | grep pr/ | xargs git branch -D; }; f"
+
+           # Get into the habit of using add --patch to create focused commits
+           ap = add --patch
+           # Amend the latest commit
+           ca = commit --amend
```
Comment (martyngigg) [inline dev-docs/source/GitConfig.rst:123]: I use
```
# Amend the latest commit reuse commit message from latest commit 
cah = commit --amend -C HEAD
```
quite a lot for fixing up the latest commit when I realise there's something I missed or I've created work-in-progress commit that I continue with later.

File: dev-docs/source/GitConfig.rst [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+           # Pretty oneline log with a bit more info than git log --oneline
+           lg = log --pretty=concise
+           # Pretty log showing only commits since master
+           lgm = log --pretty=concise master..
+           # Pretty log showing as a graph
+           lgg = log --pretty=concise --graph
```
Comment (martyngigg) [inline dev-docs/source/GitConfig.rst:130]: The only other one I use a bit is something that shows me all of the files I've changed in a branch:
```
lfc = "!f() { git log --name-only --format= master.. | sort | uniq;  }; f"
```
Feel free to ignore if you don't think it's useful.

**Review by martyngigg — COMMENTED** [review]
This is great. Hopefully a nice encouragement to use the command line. Just a few minor comments.

File: dev-docs/source/GitConfig.rst (hunk 4 above)
Comment (gemmaguest) [reply dev-docs/source/GitConfig.rst:130]: Nice :+1:

**Comment by gemmaguest** [comment]
I'd like to get this out. I can always add input from @DavidFair in a future PR. @DanNixon or @martyngigg could you approve it?

**Review by DanNixon — DISMISSED** [review]

**Review by martyngigg — DISMISSED** [review]

File: dev-docs/source/GitConfig.rst [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+.. code-block:: sh
+
+   [user]
+           # Uncomment and set these to your own details
+           #name =
+           #email =
+   [core]
+           # Set an editor you like
+           editor = vi
```
Comment (DavidFair) [inline dev-docs/source/GitConfig.rst:30]: IMO if a user is unfamiliar with git configs I would personally push them towards `nano` on Linux and `notepad++` on Windows as their default editors

File: dev-docs/source/GitConfig.rst [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+Diff/Merge Tools
+----------------
+
+You may find the following diff/merge tools helpful:
+
+.. code-block:: sh
+
+   # Uncomment lines where appropriate for specific OS
+   [diff]
+           # Set up a graphical tool for viewing diffs. I like meld because it looks good and does a good job with directory diffs.
```
Comment (DavidFair) [inline dev-docs/source/GitConfig.rst:71]: Can you add:
`but you'll need to install it separately`

**Review by DavidFair — COMMENTED** [review]
Just some minor points for less experienced developers

File: dev-docs/source/GitConfig.rst (hunk 6 above)
Comment (gemmaguest) [reply dev-docs/source/GitConfig.rst:71]: :+1: for Windows but I think meld is included on ubuntu?

Comment (DanNixon) [reply dev-docs/source/GitConfig.rst]: Can we just have some generic text that states third party tools need to be installed? Although, this does tread the line of obvious information.

**Review by martyngigg — APPROVED** [review]
Agreed. Let's get this in.

**Review by DavidFair — APPROVED** [review]
Great thanks for putting together such a comprehensive document and to everyone who provided suggestions too

**Pull request merged**